### ISSUE TITLE
Metamodel version of ProducedType.getSupertype(TypeDeclaration)

### ISSUE THREAD

**Issue by tombentley** [ceylon/ceylon.language#672]
Currently although you can ask a `ceylon.language.meta.model::Type` whether it is a subtype of another `Type`, you cannot get the prinicple instantiation of that other type that it's a subtype of. In other words we don't have the metamodel version of the typechecker's `ProducedType.getSupertype(TypeDeclaration)`.

The main impediment to resolving this is needing to implement the whole algorithm for computing principal instantiations in Ceylon. (We could implement this today for the JVM, but the metamodel needs to work on JS too).

This came up [here](https://gitter.im/ceylon/user?at=554c8436489c98827952e70a).

**Comment by gavinking** [comment]
Note: to implement this for any `Type` is perhaps a bit for work. To implement it _just_ for `ClassOrInterface` should be pretty trivial.

**Comment by gavinking** [comment]
@chochos, thoughts?

**Comment by chochos** [comment]
Right, doing it only for `ClassOrInterface` shouldn't be too hard. But is it useful?

OTOH, can we really implement the algorithm @tombentley mentions, completely in Ceylon, using the existing metamodel stuff? Perhaps if there are only a couple of things that need to be native, we can encapsule them in separate `native` methods, although I don't know if native compilation is working already for JVM.

**Comment by FroMage** [comment]
This sounds too hard, let's move it to the next release.
